Support incremental load with Arrow when cursor_column is not nullable

### dlt version

unreleased devel branch

### Describe the problem

When the cursor_column is not nullable, then the incremental loading fails with the following error:

```
E           <class 'dlt.extract.exceptions.ResourceExtractionError'>
E           In processing pipe some_data: extraction of resource some_data in transform IncrementalResourceWrapper caused an exception: cannot access local variable 'tbl_without_null' where it is not associated with a value
```

### Expected behavior

_No response_

### Steps to reproduce

```py
    data = [
        {"id": 1, "created_at": 1},
        {"id": 2, "created_at": 2},
        {"id": 3, "created_at": 2},
    ]
    schema = pa.schema([
        pa.field('id', pa.int32(), nullable=False),
        pa.field('created_at', pa.int32(), nullable=False)
    ])
    id_array = pa.array([item['id'] for item in data], type=pa.int32())
    created_at_array = pa.array([item['created_at'] for item in data], type=pa.int32())
    source_items = pa.Table.from_arrays([id_array, created_at_array], schema=schema)

    @dlt.resource
    def some_data(
        created_at=dlt.sources.incremental("created_at", on_cursor_value_missing="include")
    ):
        yield source_items

    p = dlt.pipeline(pipeline_name=uniq_id())
    p.run(some_data(), destination="duckdb")
```

### Operating system

macOS

### Runtime environment

Local

### Python version

3.11

### dlt data source

_No response_

### dlt destination

_No response_

### Other deployment details

_No response_

### Additional information

_No response_

--- a/dlt/extract/incremental/transform.py
+++ b/dlt/extract/incremental/transform.py
@@ -345,8 +345,7 @@ class ArrowIncremental(IncrementalTransform):
 
         if tbl.schema.field(cursor_path).nullable:
             tbl_without_null, tbl_with_null = self._process_null_at_cursor_path(tbl)
-
-        tbl = tbl_without_null
+            tbl = tbl_without_null
 
         # If end_value is provided, filter to include table rows that are "less" than end_value
         if self.end_value is not None:
@@ -434,11 +433,12 @@ class ArrowIncremental(IncrementalTransform):
             tbl = pyarrow.remove_columns(tbl, ["_dlt_index"])
 
         if self.on_cursor_value_missing == "include":
-            if isinstance(tbl, pa.RecordBatch):
-                assert isinstance(tbl_with_null, pa.RecordBatch)
-                tbl = pa.Table.from_batches([tbl, tbl_with_null])
-            else:
-                tbl = pa.concat_tables([tbl, tbl_with_null])
+            if tbl.schema.field(cursor_path).nullable:
+                if isinstance(tbl, pa.RecordBatch):
+                    assert isinstance(tbl_with_null, pa.RecordBatch)
+                    tbl = pa.Table.from_batches([tbl, tbl_with_null])
+                else:
+                    tbl = pa.concat_tables([tbl, tbl_with_null])
 
         if len(tbl) == 0:
             return None, start_out_of_range, end_out_of_range

--- a/tests/extract/test_incremental.py
+++ b/tests/extract/test_incremental.py
@@ -24,7 +24,6 @@ from dlt.common.utils import uniq_id, digest128, chunks
 from dlt.common.json import json
 
 from dlt.extract import DltSource
-from dlt.extract.exceptions import InvalidStepFunctionArguments
 from dlt.extract.items import ValidateItem
 from dlt.extract.resource import DltResource
 from dlt.sources.helpers.transform import take_first
@@ -914,6 +913,41 @@ def test_cursor_path_none_can_raise_on_column_missing(item_type: TestDataItemFor
         p.extract(some_data())
     assert pip_ex.value.__context__.json_path == "created_at"  # type: ignore[attr-defined]
     assert isinstance(pip_ex.value.__context__, IncrementalCursorPathMissing)
+
+
+# @pytest.mark.parametrize("item_type", ["arrow-table", "arrow-batch"])
+def test_cursor_path_not_nullable_arrow(
+    # item_type: TestDataItemFormat,
+) -> None:
+    data = [
+        {"id": 1, "created_at": 1},
+        {"id": 2, "created_at": 2},
+        {"id": 3, "created_at": 2},
+    ]
+    schema = pa.schema([
+        pa.field('id', pa.int32(), nullable=False),
+        pa.field('created_at', pa.int32(), nullable=False)
+    ])
+    id_array = pa.array([item['id'] for item in data], type=pa.int32())
+    created_at_array = pa.array([item['created_at'] for item in data], type=pa.int32())
+    source_items = pa.Table.from_arrays([id_array, created_at_array], schema=schema)
+
+    @dlt.resource
+    def some_data(
+        created_at=dlt.sources.incremental("created_at", on_cursor_value_missing="include")
+    ):
+        yield source_items
+
+    p = dlt.pipeline(pipeline_name=uniq_id())
+    p.run(some_data(), destination="duckdb")
+
+    assert_query_data(p, "select count(id) from some_data", [3])
+    assert_query_data(p, "select count(created_at) from some_data", [3])
+
+    s = p.state["sources"][p.default_schema_name]["resources"]["some_data"]["incremental"][
+        "created_at"
+    ]
+    assert s["last_value"] == 2
 
 
 def test_cursor_path_none_nested_can_raise_on_none_1() -> None:

--- a/tests/extract/test_incremental.py
+++ b/tests/extract/test_incremental.py
@@ -915,39 +915,54 @@ def test_cursor_path_none_can_raise_on_column_missing(item_type: TestDataItemFor
     assert isinstance(pip_ex.value.__context__, IncrementalCursorPathMissing)
 
 
-# @pytest.mark.parametrize("item_type", ["arrow-table", "arrow-batch"])
+@pytest.mark.parametrize("item_type", ["arrow-table", "arrow-batch"])
 def test_cursor_path_not_nullable_arrow(
-    # item_type: TestDataItemFormat,
+    item_type: TestDataItemFormat,
 ) -> None:
-    data = [
-        {"id": 1, "created_at": 1},
-        {"id": 2, "created_at": 2},
-        {"id": 3, "created_at": 2},
-    ]
-    schema = pa.schema([
-        pa.field('id', pa.int32(), nullable=False),
-        pa.field('created_at', pa.int32(), nullable=False)
-    ])
-    id_array = pa.array([item['id'] for item in data], type=pa.int32())
-    created_at_array = pa.array([item['created_at'] for item in data], type=pa.int32())
-    source_items = pa.Table.from_arrays([id_array, created_at_array], schema=schema)
 
     @dlt.resource
     def some_data(
-        created_at=dlt.sources.incremental("created_at", on_cursor_value_missing="include")
+        invocation: int,
+        created_at=dlt.sources.incremental("created_at", on_cursor_value_missing="include"),
     ):
+        if invocation == 1:
+            data = [
+                    {"id": 1, "created_at": 1},
+                    {"id": 2, "created_at": 1},
+                    {"id": 3, "created_at": 2},
+                ]
+        elif invocation == 2:
+            data = [
+                    {"id": 4, "created_at": 1},
+                    {"id": 5, "created_at": 2},
+                    {"id": 6, "created_at": 3},
+                ]
+
+        schema = pa.schema([
+            pa.field('id', pa.int32(), nullable=False),
+            pa.field('created_at', pa.int32(), nullable=False)
+        ])
+        id_array = pa.array([item['id'] for item in data], type=pa.int32())
+        created_at_array = pa.array([item['created_at'] for item in data], type=pa.int32())
+        if item_type == "arrow-table":
+            source_items = [pa.Table.from_arrays([id_array, created_at_array], schema=schema)]
+        elif item_type == "arrow-batch":
+            source_items = [pa.RecordBatch.from_arrays([id_array, created_at_array], schema=schema)]
+
         yield source_items
 
-    p = dlt.pipeline(pipeline_name=uniq_id())
-    p.run(some_data(), destination="duckdb")
 
-    assert_query_data(p, "select count(id) from some_data", [3])
-    assert_query_data(p, "select count(created_at) from some_data", [3])
+    p = dlt.pipeline(pipeline_name=uniq_id())
+    p.run(some_data(1), destination="duckdb")
+    p.run(some_data(2), destination="duckdb")
+
+    assert_query_data(p, "select id from some_data order by id", [1, 2, 3, 5, 6])
+    assert_query_data(p, "select created_at from some_data order by id", [1, 1, 2, 2, 3])
 
     s = p.state["sources"][p.default_schema_name]["resources"]["some_data"]["incremental"][
         "created_at"
     ]
-    assert s["last_value"] == 2
+    assert s["last_value"] == 3
 
 
 def test_cursor_path_none_nested_can_raise_on_none_1() -> None:

--- a/tests/extract/test_incremental.py
+++ b/tests/extract/test_incremental.py
@@ -1,52 +1,51 @@
-import os
 import asyncio
 import inspect
+import os
 import random
-from time import sleep
-from typing import Optional, Any
-from unittest import mock
 from datetime import datetime  # noqa: I251
 from itertools import chain, count
+from time import sleep
+from typing import Any, Optional
+from unittest import mock
 
 import duckdb
+import pyarrow as pa
 import pytest
 
 import dlt
+from dlt.common import Decimal
+from dlt.common.configuration import ConfigurationValueError
 from dlt.common.configuration.container import Container
 from dlt.common.configuration.exceptions import InvalidNativeValue
-from dlt.common.configuration.specs.base_configuration import configspec, BaseConfiguration
-from dlt.common.configuration import ConfigurationValueError
+from dlt.common.configuration.specs.base_configuration import (
+    BaseConfiguration,
+    configspec,
+)
+from dlt.common.json import json
 from dlt.common.pendulum import pendulum, timedelta
-from dlt.common import Decimal
 from dlt.common.pipeline import NormalizeInfo, StateInjectableContext, resource_state
 from dlt.common.schema.schema import Schema
-from dlt.common.utils import uniq_id, digest128, chunks
-from dlt.common.json import json
-
+from dlt.common.utils import chunks, digest128, uniq_id
 from dlt.extract import DltSource
-from dlt.extract.items import ValidateItem
-from dlt.extract.resource import DltResource
-from dlt.sources.helpers.transform import take_first
-from dlt.extract.incremental import IncrementalResourceWrapper, Incremental
+from dlt.extract.incremental import Incremental, IncrementalResourceWrapper
 from dlt.extract.incremental.exceptions import (
     IncrementalCursorInvalidCoercion,
+    IncrementalCursorPathHasValueNone,
     IncrementalCursorPathMissing,
     IncrementalPrimaryKeyMissing,
-    IncrementalCursorPathHasValueNone,
 )
+from dlt.extract.items import ValidateItem
+from dlt.extract.resource import DltResource
 from dlt.pipeline.exceptions import PipelineStepFailed
-
+from dlt.sources.helpers.transform import take_first
 from tests.extract.utils import AssertItems, data_item_to_list
+from tests.pipeline.utils import assert_query_data
 from tests.utils import (
+    ALL_TEST_DATA_ITEM_FORMATS,
+    TestDataItemFormat,
     data_item_length,
     data_to_item_format,
-    TestDataItemFormat,
-    ALL_TEST_DATA_ITEM_FORMATS,
 )
-
-from tests.pipeline.utils import assert_query_data
-
-import pyarrow as pa
 
 
 @pytest.fixture(autouse=True)
@@ -919,7 +918,6 @@ def test_cursor_path_none_can_raise_on_column_missing(item_type: TestDataItemFor
 def test_cursor_path_not_nullable_arrow(
     item_type: TestDataItemFormat,
 ) -> None:
-
     @dlt.resource
     def some_data(
         invocation: int,
@@ -927,30 +925,31 @@ def test_cursor_path_not_nullable_arrow(
     ):
         if invocation == 1:
             data = [
-                    {"id": 1, "created_at": 1},
-                    {"id": 2, "created_at": 1},
-                    {"id": 3, "created_at": 2},
-                ]
+                {"id": 1, "created_at": 1},
+                {"id": 2, "created_at": 1},
+                {"id": 3, "created_at": 2},
+            ]
         elif invocation == 2:
             data = [
-                    {"id": 4, "created_at": 1},
-                    {"id": 5, "created_at": 2},
-                    {"id": 6, "created_at": 3},
-                ]
+                {"id": 4, "created_at": 1},
+                {"id": 5, "created_at": 2},
+                {"id": 6, "created_at": 3},
+            ]
 
-        schema = pa.schema([
-            pa.field('id', pa.int32(), nullable=False),
-            pa.field('created_at', pa.int32(), nullable=False)
-        ])
-        id_array = pa.array([item['id'] for item in data], type=pa.int32())
-        created_at_array = pa.array([item['created_at'] for item in data], type=pa.int32())
+        schema = pa.schema(
+            [
+                pa.field("id", pa.int32(), nullable=False),
+                pa.field("created_at", pa.int32(), nullable=False),
+            ]
+        )
+        id_array = pa.array([item["id"] for item in data], type=pa.int32())
+        created_at_array = pa.array([item["created_at"] for item in data], type=pa.int32())
         if item_type == "arrow-table":
             source_items = [pa.Table.from_arrays([id_array, created_at_array], schema=schema)]
         elif item_type == "arrow-batch":
             source_items = [pa.RecordBatch.from_arrays([id_array, created_at_array], schema=schema)]
 
         yield source_items
-
 
     p = dlt.pipeline(pipeline_name=uniq_id())
     p.run(some_data(1), destination="duckdb")
@@ -2523,7 +2522,7 @@ def test_allow_external_schedulers(item_type: TestDataItemFormat) -> None:
 
 @pytest.mark.parametrize("yield_pydantic", (True, False))
 def test_pydantic_columns_validator(yield_pydantic: bool) -> None:
-    from pydantic import BaseModel, Field, ConfigDict
+    from pydantic import BaseModel, ConfigDict, Field
 
     # forbid extra fields so "id" in json is not a valid field BUT
     # add alias for id_ that will serde "id" correctly

--- a/tests/utils.py
+++ b/tests/utils.py
@@ -3,7 +3,7 @@ import os
 import platform
 import sys
 from os import environ
-from typing import Any, Iterable, Iterator, List, Literal, Union, get_args
+from typing import Any, Iterable, Iterator, Literal, Union, get_args
 from unittest.mock import patch
 
 import pytest
@@ -18,7 +18,7 @@ from dlt.common.configuration.specs import RunConfiguration
 from dlt.common.configuration.specs.config_providers_context import (
     ConfigProvidersContext,
 )
-from dlt.common.pipeline import PipelineContext
+from dlt.common.pipeline import PipelineContext, SupportsPipeline
 from dlt.common.runtime.init import init_logging
 from dlt.common.runtime.telemetry import start_telemetry, stop_telemetry
 from dlt.common.schema import Schema
@@ -26,7 +26,6 @@ from dlt.common.storages import FileStorage
 from dlt.common.storages.versioned_storage import VersionedStorage
 from dlt.common.typing import StrAny, TDataItem
 from dlt.common.utils import custom_environ, uniq_id
-from dlt.common.pipeline import PipelineContext, SupportsPipeline
 
 TEST_STORAGE_ROOT = "_storage"
 

